debian/libsites.1
.TH "LIBSITES" "1" "" "MANUAL" ""
.SH NAME
.PP
libsites \- GBV Standortverzeichnis
.SH NAME
.PP
libsites \- GBV\-Standortverzeichnis
.SH SYNOPSIS
.PP
The application is automatically started as service, listening on port
6013
.IP
.nf
\f[C]
sudo\ service\ libsites\ {status|start|stop|restart}
\f[]
.fi
.SH DESCRIPTION
.PP
libsites implements an RDF\-based web registry of library locations.
.SH INSTALLATION
.PP
The software is released as Debian package for Ubuntu 14.04 LTS.
Other Debian based distributions \f[I]might\f[] work too.
Releases can be found at <https://github.com/gbv/libsites/releases>
.PP
To install required dependencies either use a package manager such as
\f[C]gdebi\f[], manually install dependencies (inspectable via
\f[C]dpkg\ \-I\ libsites_*.deb\f[]):
.IP
.nf
\f[C]
sudo\ dpkg\ \-i\ ...\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ #\ install\ dependencies
sudo\ dpkg\ \-i\ libsites_X.Y.Z_amd64.deb\ \ \ \ #\ change\ X.Y.Z
\f[]
.fi
.PP
After installation the service is available at localhost on port 6013.
Better put the service behind a reverse proxy to enable SSL and nice
URLs!
.SH ADMINISTRATION
.SS Configuration
.PP
Config file \f[C]/etc/default/libsites\f[] only contains basic server
configuration in form of simple key\-values pairs:
.IP \[bu] 2
\f[C]PORT\f[] \- port number (required, 6013 by default)
.IP \[bu] 2
\f[C]WORKERS\f[] \- number of parallel connections (required, 5 by
default).
.PP
The actual content is retrieved from the German ISIL directory (hosted
by ZDB) and the git repository <https://github.com/gbv/libsites-config>.
.SS Updates of content
.PP
Ein Update läuft täglich sowie (ohne ZDB\-Update) bei Benachrichtigung
per GitHub Webhook an den URL\-Pfad \f[C]/update\f[].
Der Verlauf des Update\-Skripts wird in
\f[C]/var/log/libsites/update.log\f[] festgehalten.
Zur Not kann das Update auch manuell angestoßen werden:
.IP
.nf
\f[C]
sudo\ \-u\ libsites\ /etc/cron.daily/libsites
\f[]
.fi
.SS Logging
.PP
Log files are written in \f[C]/var/log/libsites/\f[] and kept for 30 day
by default:
.IP \[bu] 2
\f[C]access.log\f[] \- HTTP request and responses in Apache Log Format
.IP \[bu] 2
\f[C]server.log\f[] \- Web server messages (when server was started and
stopped)
.IP \[bu] 2
\f[C]update.log\f[] \- Update script output
.SH SEE ALSO
.PP
The source code of libsites is managed in a public git repository at
<https://github.com/gbv/libsites>.
Please report bugs and feature request at
<https://github.com/gbv/libsites/issues>!
.PP
The Changelog is located in file \f[C]debian/changelog\f[].
.PP
Development guidelines are given in file \f[C]CONTRIBUTING.md\f[].


debian/changelog
libsites (0.2.1) unstable; urgency=medium

  * Allow GitHub webhooks from anywhere
  * Fixed isil.csv update bug

 -- Jakob Voß <jakob.voss@gbv.de>  Wed, 09 Sep 2015 15:10:31 +0200

libsites (0.2.0) unstable; urgency=medium

  * refactored update

 -- Jakob Voß <jakob.voss@gbv.de>  Tue, 08 Sep 2015 14:01:31 +0200

libsites (0.1.5) unstable; urgency=medium

  * migrated packaging to Debian

 -- Jakob Voß <jakob.voss@gbv.de>  Thu, 23 Jun 2015 10:00:00 +0100


lib/GBV/App/Libsites.pm
package GBV::App::Libsites;

use v5.14.2;
no warnings 'experimental::smartmatch';

use parent 'Plack::Component';
use Plack::Util::Accessor qw(config isil app tt);
use Plack::Builder;

use Log::Contextual qw(:log :dlog);
use Plack::Middleware::TemplateToolkit;
use Plack::App::Directory::Template;
use Plack::App::RDF::Files;
use RDF::Trine::Model;
use RDF::Trine::Parser;
use Time::Piece;
use List::Util qw(max);
use RDF::NS;
use RDF::Lazy;
use constant NS => RDF::NS->new();

our $VERSION="0.2.1";

sub prepare_app { # sub BUILD
    my ($self) = @_;

    # TODO: 'with
    for ($ENV{LIBSITES_CONFIG}, './libsites-config', '/etc/libsites') {
        $self->config($_);
        last if -d $_;
    }

    my $isildir = $self->config . '/isil';
    mkdir $isildir unless -d $isildir;

    my $tt = Plack::Middleware::TemplateToolkit->new( 
        INCLUDE_PATH => 'public',
        INTERPOLATE  => 1, 
        VARIABLES    => { base => './' },
        vars         => { formats => [qw(ttl rdfxml nt json)] },
        404          => '404.html', 
        500          => '500.html',
        PRE_PROCESS  => 'header.html',
        POST_PROCESS => 'footer.html',
        dir_index    => 'about.html',
    );
    $tt->prepare_app;

    my $app = builder {
        enable 'Static', 
            root => 'public', 
            path => qr{\.(css|png|gif|js|ico)$};
        enable 'Static',
            root => $self->config,
            path => qr{libsites\.ttl$};
        enable 'JSONP';
        enable 'Negotiate',
            parameter => 'format',
            extension => 0,
            formats => {
                ttl    => { type => 'text/turtle' },
                rdfxml => { type => 'application/rdf+xml' },
                nt     => { type => 'text/plain' },
                n3     => { type => 'text/n3' },
                json   => { type => 'application/rdf+json' },
                html   => { type => 'text/html' },
                _      => { charset => 'utf-8', }
            };
        builder {
            mount '/source' => Plack::App::Directory::Template->new(
                    root         => $isildir,
                    VARIABLES    => { base => '../../' }, # TODO: dynamic
                    templates    => 'public',
                    INTERPOLATE  => 1, 
                    PRE_PROCESS  => 'header.html',
                    POST_PROCESS => 'footer.html',
                );
            mount '/isil' => builder {
                enable_if { $_[0]->{'negotiate.format'} eq 'html' } sub {
                    my ($app) = @_;
                    sub {
                        my ($env) = @_;
                        $env->{'psgi.streaming'} = 1;
                        $env->{'negotiate.format'} = 'nt';
                        my $res = $app->($env);

                        Plack::Util::response_cb( $res, sub {
                            my $res = shift;
                            my $uri = $env->{'rdf.uri'};
                            my $iter = $env->{'rdf.iterator'};
                            my $model = RDF::Trine::Model->new;
                            if ($res->[0] eq '200' and $iter) {
                                $model->add_iterator($iter);
                            }
                            my $res2 = $self->call_html($env,$model);
                            $res->[$_] = $res2->[$_] for (0..2);
                        }); 
                    }
                };
                Plack::App::RDF::Files->new( 
                    base_dir => $isildir,
                    base_uri => 'http://uri.gbv.de/organization/isil/',
                    path_map => sub { $_ = shift; $_ =~ s/\@.+$//; $_ },
                    normalize => 'NFC',
                )->to_app;
            };
            mount '/' => $tt;
        };
    };

    $self->tt($tt);
    $self->app($app);
}

sub call_html {
    my ($self, $env, $rdf) = @_;

    my $uri = $env->{'rdf.uri'};
    my $lazy = RDF::Lazy->new( $rdf, namespaces => NS );

    $env->{'tt.vars'}->{uri} = $lazy->resource($uri);
    $env->{'tt.vars'}->{javascript} = [ 'OpenLayers.js' ];
    add_env_sources($env, $uri);
    $env->{'tt.path'} = '/organization.html';   
    $env->{'tt.vars'}{base} = '../';

    $self->tt->call($env);
}
 
sub add_env_sources {
    my ($env, $uri) = @_;
    my $rdffiles = $env->{'rdf.files'};

    my $isil = ($uri =~ /http:\/\/uri.gbv.de\/organization\/isil\/(.+)/ ? $1 : '');
    $isil =~ s/@.*//;

    my @sources; 
    foreach my $file ( keys %$rdffiles ) {
        my $about = $rdffiles->{$file};
        my $s = $file; $s =~ s/\..+$//;
        my %h = (%$about, file => $file);
        given ($s) {
            when('sites') {
                $h{title} = 'GBV Standortverzeichnis';
                $h{href} = "https://github.com/gbv/libsites-config/blob/master/isil/$isil/sites.txt";
            };
            when('opac') {
                $h{'title'} = 'GBV Datenbankverzeichnis';
                $h{'href'} = 'http://uri.gbv.de/database/opac-'.lc($isil);
            };
            when('lobid') { 
                $h{title} = 'lobid.org';
                $h{href} = "http://lobid.org/organisation/$isil";
            };
            when('zdb') {
                $h{title} = 'Deutsches ISIL- und Sigelverzeichnis (PICA)';
                $h{href} = "http://ld.zdb-services.de/resource/organisations/$isil";
            };
            when('zdbrdf') {
                $h{title} = 'Deutsches ISIL- und Sigelverzeichnis (RDF)';
                $h{href} = "http://ld.zdb-services.de/resource/organisations/$isil.rdf";
            };
        }
        push @sources,\%h;
    }

    $env->{'tt.vars'}{sourcedir} = $isil; 
    $env->{'tt.vars'}{sources} = \@sources;
    $env->{'tt.vars'}{timestamp} = localtime( max map { $_->{mtime} } values %$rdffiles )->strftime;

    return @sources; 
}

sub call {
    my ($self, $env) = @_;
    $self->app->($env);
}

1;
